flashlight/fl/tensor/backend/jit/JitBackend.cpp

```diff
@@ -31,10 +32,11 @@ TensorBackendType JitBackend::backendType() const {
 
 /* -------------------------- Compute Functions -------------------------- */
 
-void JitBackend::eval(const Tensor& /* tensor */) {
-  // Launch computation for a given tensor. Can be a noop for non-async
-  // runtimes.
-  FL_JIT_BACKEND_UNIMPLEMENTED;
+void JitBackend::eval(const Tensor& tensor) {
+  auto& jitTensor = toJitTensorBase(tensor);
+  assert(&jitTensor.backend() == this);
+  jitTensor.eval();
```

Comment:
This is interesting -- the separation of concerns here is a bit weird. Should the JIT wrap that call to eval for the most part and basically emit ops that another backend has to execute?

I can see a scenario in which the "deep" eval ends up doing excess computation, but maybe I'm thinking about this wrong

thoughts?

Reply:
> the separation of concerns here is a bit weird

Agree. So the whole JitTensor backend serves as both **parser** and **driver** (mostly the former, but given the Tensor API, it also needs to be the one that kicks off the evaluation, e.g., upon `tensor.toString()` call.

> Should the JIT wrap that call to eval for the most part and basically emit ops that another backend has to execute?

Not sure if I fully understand the question. But `emit ops` sounds like the process of building up the JIT nodes/graph? I decided to put the `jitTensor.eval()`logic (launching Evaluator) inside `JitTensorBase`, because the it serves as the driver logic that'll call optimizer and update the node as well (in upcoming PR). It's slightly more indirect to update the node field from `JitBackend`.

> I can see a scenario in which the "deep" eval ends up doing excess computation, but maybe I'm thinking about this wrong

So the motivation for this was that for things like `JitTensor<JitTensor<T>>`, a single `fl::eval` won't kick off computation in `T`. But I don't have any other example in mind and this is unlikely to be used in practice:).

I guess it really depends on the desired semantics for `fl::eval` for Tensor clients, let me know what you think.

Reply:
Yep, forgot it was the **driver** idea; vis-a-vis this abstraction the JIT would be constructing a modified graph vis a vis the wrapped backend. Makes perfect sense 👍